src/Microsoft.AspNetCore.SystemWebAdapters/Adapters/SystemWebAdaptersExtensions.Framework.cs

```diff
@@ -18,45 +20,56 @@ public static ISystemWebAdapterBuilder AddSystemWebAdapters(this HttpApplication
             throw new ArgumentNullException(nameof(state));
         }
 
-        if (state[Key] is not ISystemWebAdapterBuilder builder)
+        if (state[BuilderKey] is not ISystemWebAdapterBuilder builder)
         {
-            builder = new Builder();
-            state[Key] = builder;
+            builder = new SystemWebAdapterBuilder(new ServiceCollection());
+            state[BuilderKey] = builder;
         }
 
         return builder;
     }
 
     public static ISystemWebAdapterBuilder AddProxySupport(this ISystemWebAdapterBuilder builder, Action<ProxyOptions> configure)
     {
+        if (builder is null)
+        {
+            throw new ArgumentNullException(nameof(builder));
+        }
+
         if (configure is null)
         {
             throw new ArgumentNullException(nameof(configure));
         }
 
-        return builder.AddModule(configure, static options => new ProxyHeaderModule(options));
-    }
-
-    internal static ISystemWebAdapterBuilder? GetSystemWebBuilder(this HttpApplicationState state)
-        => state[Key] as ISystemWebAdapterBuilder;
-
-    private static ISystemWebAdapterBuilder AddModule<TOptions>(this ISystemWebAdapterBuilder builder, Action<TOptions> configure, Func<TOptions, IHttpModule> factory)
-        where TOptions : class, new()
-    {
-        var options = new TOptions();
-        configure(options);
-
-        return builder.AddModule(factory(options));
-    }
+        builder.Services.AddScoped<IHttpModule, ProxyHeaderModule>();
+        builder.Services.AddOptions<ProxyOptions>()
+            .Configure(configure);
 
-    private static ISystemWebAdapterBuilder AddModule(this ISystemWebAdapterBuilder builder, IHttpModule module)
-    {
-        builder.Modules.Add(module);
         return builder;
     }
 
-    private class Builder : ISystemWebAdapterBuilder
+    internal static IServiceProvider? GetServiceProvider(this HttpApplicationState state)
     {
-        public ICollection<IHttpModule> Modules { get; } = new List<IHttpModule>();
+        var serviceProvider = state[ServicesKey] as IServiceProvider;
```

Comment:
how should we handle disposing the rooted container?

Reply:
Maybe DomainUnload in case anyone is expecting Dispose to be called to write persist any state correctly or dispose pattern isn't correctly handled.

Reply:
Actually, looking at this a bit more I think a better option would be to use `HttpApplication`. If we make the `AddSystemWebAdapters` extension take the application rather than the application state, then we can both register the service collection in the state and use HttpApplication's `Disposed` event to clean up the provider (if one has been created) at app shutdown. This looks like a more correct way to clean up application-scoped resources.